### DISCUSSION TITLE
Display an arbitrary 3D mesh

### DISCUSSION THREAD

**Discussion by tbarusseau** [emilk/egui#643] [Q&A] [answered]
Hello! I'm using `egui` to make a 3D-model viewer, and am wondering how to display a 3D mesh.

I can see in the documentation that I can do `ui.painter().add(egui::Shape::Mesh(/* Some mesh */))`.

Mesh needs an indices array for triangles (which doesn't pose much issues), but also a [Vertex][1] array. However it's indicated that this is "the 2D vertex type", implying that there might be a 3D vertex type?

If there isn't, do I need to project each of my 3D-model's vertex to egui's 2D frame? Manually doing occlusion culling etc?

Thanks in advance.

[1]: https://docs.rs/epaint/0.13.0/epaint/struct.Vertex.html

**Comment by emilk** [comment, marked as answer]
3D is outside the scope of egui. You can project to 2D and do your own culling, but that is tedious.

I suggest you use something like [macroquad](https://github.com/not-fl3/macroquad) (with [egui-macroquad](https://github.com/optozorax/egui-macroquad)) and render your mesh to a texture and display it with [`egui::Image`](https://docs.rs/egui/0.13.1/egui/widgets/struct.Image.html).

**Reply by tbarusseau** [reply]
Thank you! Also thanks for having the discussions enabled.

Cheers and good luck for the future of this awesome project!
